### PR TITLE
Interval set type.

### LINKED ISSUE #270: Interval set type
We need:
* multiply interval vs interval
* multiply interval vs scalar
* take Minkowski sum of interval and interval
* `overapproximate` (no-op)
* `decompose`

### PR DESCRIPTION
**Pull request by mforets**
Closes #270.

### CODE REVIEW

File: src/Interval.jl [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+- `d` -- direction
+- `x` -- interval
+
+### Output
+
+Support vector in the given direction.
+"""
+function σ(d::AbstractVector{N}, x::Interval{N, IN})::AbstractVector{N} where {N, IN <: IA.AbstractInterval{N}}
+    return d[1] > 0 ? [x.dat.hi] : [x.dat.lo]
```
[the remark this answers is not in the document]
Comment (mforets) [reply src/Interval.jl]: do you mean that we don't check dimensions? it's not yet global, but yes, some types have an assertion.

```julia
julia> B
LazySets.Ball1{Float64}([0.0, 0.0], 1.0)

julia> σ(rand(4), B) # support vector of dim 4 over a dim 2 set
2-element Array{Float64,1}:
 0.0
 1.0
```

**Comment by mforets** [comment]
actually, for consistency with all other symbolic set types, `+` should trigger a fresh `MinkowskiSum` instance. `minkowski_sum` should behave like my current `+`. thoughts?

**Comment by schillic** [comment]
> actually, for consistency with all other symbolic set types, + should trigger a fresh MinkowskiSum instance.

Okay, or make a difference between `⊕` and `+`, like for `×` and `*`.

File: src/Interval.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (schillic) [reply src/Interval.jl]: Yes, that is what I meant.

**Comment by mforets** [comment]
.. yes `*` should be a new `CartesianProduct`.  

then i take back my suggestion and propose that for intervals, `+` and `*` are NOT symbolic.

**Comment by schillic** [comment]
I am fine with either way.

File: test/unit_Interval.jl [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+m = x + y
+@test dim(m) == 1
+@test σ([1.0], m) == [1.5]
+@test σ([-1.0], m) == [-2.0]
+
+# Product of intervals: uses the * operator
+p = x * y
+@test dim(p) == 1
+@test σ([1.0], p) == [0.5]
+@test σ([-1.0], p) == [-2.0]
```
[the remark this answers is not in the document]
Comment (mforets) [reply test/unit_Interval.jl]: done. any other method that we would like to have?

[the remark this answers is not in the document]
Comment (schillic) [reply test/unit_Interval.jl]: Those are all functions that are defined here :+1:

File: test/unit_Interval.jl [hunk 3]
```diff
@@ -11,16 +11,44 @@ x = LazySets.Interval(0.0, 1.0)
...
 @test σ([-1.0], p) == [-2.0]
+v = vertices_list(p)
+@test [0.5] in v && [-2.0] in v
+
+# test different arithmetic operations
+r = (x + y) - (d + p)
+@test low(r) == -5.5 && high(r) == 4.0
+
+# Minkowski sum
+
```
Comment (schillic) [inline test/unit_Interval.jl]: Something missing?

Comment (mforets) [reply test/unit_Interval.jl]: #278 and the unit test :)

File: test/unit_Interval.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mforets) [reply test/unit_Interval.jl]: i really meant: do we want some other new method for the `Interval` type? i can think of `overapproximate`.

File: test/unit_Interval.jl (hunk 3 above)
Comment (mforets) [reply test/unit_Interval.jl]: done.

File: test/unit_Interval.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mforets) [reply test/unit_Interval.jl]: there is also `∈` and `⊆`.

[the remark this answers is not in the document]
Comment (schillic) [reply test/unit_Interval.jl]: Ah yes, we definitely want to have `overapproximate` to be a no-op.

**Comment by mforets** [comment]
just saw that `center(x::Interval)` should return a vector (not a scalar), so that `Interval` works smoothly with other interface methods, eg. `an_element(x)`.

File: test/unit_Interval.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (schillic) [reply test/unit_Interval.jl]:
```julia
function ∈(x::AbstractVector{N}, I::LazySets.Interval{N, IN}) where ...
    @assert length(x) == dim(I)
    return low(I) <= x[1] <= high(I)
end
```

[the remark this answers is not in the document]
Comment (schillic) [reply test/unit_Interval.jl]:
```julia
function ⊆(X::LazySet{N}, I::LazySets.Interval{N, IN}) where ...
    @assert dim(X) == dim(I)
    return low(I) <= σ(X, [-one(N)]) && σ(X, [one(N)]) <= high(I)
end
function ⊆(I::LazySets.Interval{N, IN}, X::LazySet{N}) where ...
    @assert dim(X) == dim(I)
    return σ(X, [-one(N)]) <= low(I) && high(I) <= σ(X, [one(N)])
end
function ⊆(I1::LazySets.Interval{N, IN}, I2::LazySets.Interval{N, IN}) where ...
    return low(I2) <= low(I1) && high(I2) <= high(l1)
end
```
(no witness production)

[the remark this answers is not in the document]
Comment (mforets) [reply test/unit_Interval.jl]: that is all good. while we don't extend for witness production i've just called the interval arithmetic backend for these.

**Comment by mforets** [comment]
having added the new `overapproximate(S::LazySets.Interval{N, IN}, ::Type{Hyperrectangle}) where {N, IN <: IntervalArithmetic.AbstractInterval{N}}` i stumble upon some compilation error and i'm clueless. in a fresh session:

```julia
julia> using LazySets # so far so good
INFO: Precompiling module LazySets.
julia> using IntervalArithmetic  # boom
WARNING: Error requiring IntervalArithmetic from LazySets.Approximations:
UndefVarError: Interval not defined
Stacktrace:
 [1] err(::LazySets.Approximations.##6#13, ::Module, ::String) at /Users/forets/.julia/v0.6/Requires/src/require.jl:42
 [2] withpath(::LazySets.Approximations.##5#12, ::String) at /Users/forets/.julia/v0.6/Requires/src/require.jl:32
 [3] (::LazySets.Approximations.##4#11)() at /Users/forets/.julia/v0.6/Requires/src/require.jl:53
 [4] _collect(::Array{Function,1}, ::Base.Generator{Array{Function,1},Requires.##3#4}, ::Base.EltypeUnknown, ::Base.HasShape) at ./array.jl:488
 [5] loadmod(::Symbol) at /Users/forets/.julia/v0.6/Requires/src/require.jl:20
 [6] require(::Symbol) at ./loading.jl:408
 [7] eval(::Module, ::Any) at ./boot.jl:235
 [8] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
 [9] macro expansion at ./REPL.jl:97 [inlined]
 [10] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
```

**Comment by schillic** [comment]
Confirmed here. Replacing `@require IntervalArithmetic begin [...] end` by `using IntervalArithmetic` solves the error, but we want to use `@require`.
My guess was that `Interval` is not exported correctly (the `export` is only inside the `@reexport`). But adding `export Interval` outside did not help.

**Comment by schillic** [comment]
The problem is that `Requires` stores the things to execute by the module name and in Julia `LazySets.Approximations` comes before `LazySets`.

You can wrap the `Interval.jl` inside a new module that comes before `Approximations`, e.g., `module a`. But then you always need to write `LazySets.a.Interval` and you have to spell out `IA`.

**Pull request merged**